cfpb/cfgov-refresh
# Fixes DF-1588 - implements Browserify JS module loader

Refactors codebase to use Browserify for JS modules.
Fixes https://github.com/cfpb/cfgov-refresh/issues/244

The general idea is this, all code is broken across three types of modules:
 - **component-specific** =  These appear in `/src/static/js/modules/`. These work on a part of a page and are aggregated together in page-specific modules (see next).
 - **page-specific** = These appear in `/src/static/js/routes/`. These files match the URL structure of the page that they get attached to (which unfortunately means many are named `index.js`, but it has to be that way as far as I can see to keep a clean link to actual page URLs). 
 - **site-wide** = This file appears in `/src/static/js/routes/common.js`. It is added to all pages.

> **Note:** New files can be added to the routes (page-specific modules) and they will automatically be included in Browserify without any further configuration. This works through the following:
- The `browserifyPageScripts` function in the Gruntfile looks at the directory structure of the source and in addition to grabbing all the page-specific modules, generates a `routes-manifest.json` file that includes all the pages that have associated page-specific scripts.
- The routes manifest is embedded in the bottom of the base jinja template and is dynamically injected with the current page path being viewed.
- The `common.js` site-wide JS file reads the JSON from the routes manifest and if the current page path has an associated entry for a JS module it dynamically loads the module via injecting a script tag in the head of the page. This prevents all pages from requesting a page-specific module, even if they don't have one (which would throw a 404 in the console).

## Additions

- Adds `/src` directory for JS (and eventually other assets) source code.
- Adds `browserify` and dependencies.
- Adds `browserifyPageScripts` helper function to Gruntfile so that files to process can be pulled from disk [instead of being manually entered](https://github.com/cfpb/owning-a-home/blob/master/Gruntfile.js#L130-L149).
- Adds `/static/js/routes-manifest.json` to track page-specific scripts that exist.
- Adds `/modules/util/web-storage-proxy.js` utility module for handling local and session storage.
- Adds `/modules/util/js-loader.js` utility module for dynamically loading JS scripts at runtime.
- Adds additional ESLint option flags in `space-in-brackets` rule.

## Removals

- Removes inline JS from footer of homepage, `the-bureau/bureau-structure/index.html` and `the-bureau/index.html`.

## Changes

- Updates `grunt` to `~0.4.5`.
- Updates `jquery` to `^1.11.3`.
- Replaces `grunt-uglify` with `uglifyify`.
- Breaks JS into Browserify modules.
- Replaces `uglify` with `lintjs` in `jsdev` task.
- Changes ESLint indent amount to 2 spaces to match CFPB front-end standards.
- Turns off ESLint `func-names` setting because its too verbose for the gain it provides.
- General code cleanup, fixing ESLint violations (https://github.com/cfpb/cfgov-refresh/issues/300), removing extraneous whitespace, etc.

## Testing

- Throw out `/node_modules` and `/vendor` and run `npm install` and `grunt vendor`.
- Run `grunt browserify:build`.
- Check console in browser for errors.

## Review

- @jimmynotjim 
- @sebworks 
- @ascott1 
- @cfarm
- @KimberlyMunoz 

## Preview

Should be no change. If there's a change, something is broken.

## Notes

- TODO: Check if all the JS code is actually in use.
- TODO: Check if some of the global common code can move to pages.
- TODO: Add JS tests.
- `usebanner:js` in Gruntfile is broken. I'm not sure what we even use that for.
- The generated and minified JS is in `/static/js/`, once https://github.com/cfpb/cfgov-refresh/pull/466 moves forward it should be moved to `/dist/` and be added to the gitignore, because it currently contains my local path values in the source.
- I wrapped the `common.js` code in a jQuery doc ready function, testing would be helpful to determine if this isn't necessary since the code is in the footer. @Scotchester suggested there might be cases where it may still run before the DOM is ready.
- The Google Tag Manager (GTM) contains jQuery code, but loads at the top of the body, while the rest of the code (including loading jQuery) happens at the bottom. This originally was throwing errors for me but stopped when a brought the Browserify code together. The current setup of script location matches Owning a Home. However, I believe this may be hiding a bug. I think what's happening is that since the GTM is fetching its code from a CDN and the rest of the code (including jQuery) is being fetched locally, the scripts at the bottom of the body are running before the GTM has fetched its code, so when it does initialize, jQuery is available. I suspect under certain conditions it could initialize before the footer code, leading to an error of jQuery not being found.
